hoptimator-catalog/src/main/java/com/linkedin/hoptimator/catalog/Resource.java

```diff
@@ -182,31 +186,38 @@ public SimpleEnvironment with(String key, String value) {
     }
 
     @Override
-    public String get(String key) {
-      if (!vars.containsKey(key)) {
+    public String getOrDefault(String key, String defaultValue) {
+      if (defaultValue == null && !vars.containsKey(key)) {
```

Comment:
what happens if the `defaultValue` is an empty or whitespace only string?

Reply:
It should get rendered by the template as a blank, which may actually be what you want in some cases, I guess.

I supposed you could even use the string `"null"`, e.g. you could have the hint `subscription.spec.hints.kafka.numPartitions: "null"` s.t. the template renders `numPartitions: null`, which the Kafka controller interprets as "whatever the broker has configured".

Notice hints are just strings, so we should accept `""`, `"null"`, or whatever the user wants to inject into the template.

Reply:
(...and the same with `defaultValue` ... should be whatever raw string the template provides.)

Reply:
thinking more, what if someone wants to provide more metadata in hints? don't we need to make it object? do we need to namespace by the adapter type? like if you say database is queueing , hints object needs to have Kafka/stream related metadata Vs database=brooklin, hints need to have Brooklin datastream related hints?

Reply:
Yes, I wonder about this too. Instead of `kafka.numPartitions` we could have e.g. `QUEUING.numPartitions`, which would solve some problems I mentioned above to @vmaheshw. Specifically, the current implementation here applies to _all_ Kafka topics in a pipeline, which _probably_ isn't what you want. If you have a pipeline that reads from a Kafka topic and writes to another, the current implementation will try to add partitions to the sink topic _and the source topic_, which is a bit weird.

If instead we had e.g. `QUEUING.numPartitions` vs `TRACKING.numPartitions`, then we could explicitly control the partitions for the source and sink independently _iff_ they happen to be in different databases. If the same database is used as both source and sink (e.g. QUEUING --> QUEUING), this doesn't really help us! So I don't think using the database name is a perfect solution either.

We could maybe have hints like `QUEUING.my-topic.numPartitions`, i.e. we could scope hints down to the specific table in a specific database. That would solve the problems above, but it might be hard to provide hints to internal tables. For example, if we have a CDC stream as part of the pipeline, we'd maybe have something like:

```
spec:
   sql: SELECT * FROM MYSQL.MYTABLE
   hints:
     CDC.MYTABLE_CHANGES.numPartitions: "2"
```

... which works fine but means that the user must know about the internal table `CDC.MYTABLE_CHANGES`.

I like your idea of making these objects instead of just strings. What about smth like:

````
hints:
   - database: QUEUING
      table: my_topic
      key: numPartitions
      value: "4"
```

... where the database and table properties are optional.

If that makes sense, lets add support for the global scope (minus database and table scope) for now, and it will be easy to add the database and table scope later.

Reply:
We talked out-of-band and decided that the most common (and really, the only logical) use-case for hints is to control resources related to _sinks_. So instead of enabling hints to be scoped to arbitrary resources, we decided to limit them just to sinks, i.e. all the resources related to the sink table.